File: backend/risk_engine.py
# ==============================================================================
# Fraud Detection and Risk Assessment System
# Explainable AI (XAI) — 12-Pattern Behavioral Risk Engine
# ==============================================================================
from typing import Dict, Any, List, Optional
from datetime import datetime

RISK_THRESHOLDS = {
    "LOW": (0, 29),
    "MEDIUM": (30, 59),
    "HIGH": (60, 79),
    "CRITICAL": (80, 100),
}

PATTERN_DESCRIPTIONS = {
    "P1":  "Repeated Same-Amount Transactions (Rapid)",
    "P2":  "Doubling/Multiplying Amount to Same Recipient",
    "P3":  "Unusual Transaction Frequency Burst",
    "P4":  "Amount Outside Historical Range",
    "P5":  "New Recipient + Unusual High Amount",
    "P6":  "New Recipient + Multiple Rapid Transactions",
    "P7":  "Multiple Different Recipients in Short Period",
    "P8":  "Unusual Time of Day / Late Night",
    "P9":  "Round / Structured Amount to New Recipient",
    "P10": "Device Anomaly (Unrecognized Device)",
    "P11": "Geographic / Location Anomaly",
    "P12": "Rapid Large-Value Transaction Spike",
}


def get_risk_level(score: int) -> Dict[str, Any]:
    s = max(0, min(100, int(score)))
    if s < 30:
        return {"level": "LOW",      "label": "Low Risk",      "badge": "badge-low",      "color": "#10b981"}
    elif s < 60:
        return {"level": "MEDIUM",   "label": "Medium Risk",   "badge": "badge-medium",   "color": "#f59e0b"}
    elif s < 80:
        return {"level": "HIGH",     "label": "High Risk",     "badge": "badge-high",     "color": "#ef4444"}
    else:
        return {"level": "CRITICAL", "label": "Critical Risk", "badge": "badge-critical", "color": "#dc2626"}


def _parse_hour(time_str: str, hour_hint=None) -> int:
    if hour_hint is not None:
        return int(hour_hint)
    try:
        ts = str(time_str).strip()
        if "AM" in ts.upper() or "PM" in ts.upper():
            return datetime.strptime(ts, "%I:%M %p").hour
        return int(ts.split(":")[0])
    except Exception:
        return 12


def evaluate_transaction_risk(
    transaction: Dict[str, Any],
    profile: Optional[Dict[str, Any]] = None,
    recent_transactions: Optional[List[Dict[str, Any]]] = None
) -> Dict[str, Any]:
    """
    Evaluates a UPI transaction against the user's behavioral profile
    using 12 configurable fraud/suspicion patterns (P1-P12).
    """
    amount = float(transaction.get("amount", 0))
    receiver = str(transaction.get("receiver", "")).strip()
    receiver_upi = str(
        transaction.get("receiver_upi") or transaction.get("receiverUpi") or ""
    ).strip().lower()
    time_str = str(transaction.get("time", "12:00 PM"))
    device = str(transaction.get("device") or "")
    location = str(transaction.get("location") or "")

    hour = _parse_hour(time_str, transaction.get("hour"))
    recent = recent_transactions or []

    factors: List[Dict[str, Any]] = []
    detected_patterns: List[str] = []

    # COLD START
    if not profile or profile.get("totalCount", 0) == 0:
        if amount > 20000:
            factors.append({
                "name": "High Value Transaction (Unprofiled)",
                "contribution": 35,
                "pattern": "P4",
                "description": f"Rs.{amount:,.2f} is an elevated amount with no established spending baseline."
            })
            detected_patterns.append("P4")
        else:
            factors.append({
                "name": "Baseline Verification",
                "contribution": 10,
                "pattern": "--",
                "description": "Transaction evaluated under initial cold-start parameters (no history yet)."
            })

        if hour < 5 or hour >= 23:
            factors.append({
                "name": "Late Night Activity",
                "contribution": 20,
                "pattern": "P8",
                "description": f"Transaction initiated at {time_str} -- outside normal active hours."
            })
            detected_patterns.append("P8")

        raw_score = sum(f["contribution"] for f in factors)
        score = max(5, min(95, raw_score))
        risk_info = get_risk_level(score)
        return _build_result(score, risk_info, detected_patterns, factors, "Cold Start (No Profile)")

    # Profile parameters
    avg_amt = float(profile.get("avgAmount", 1500))
    max_amt = float(profile.get("maxAmount", 10000))
    std_dev = float(profile.get("stdDev", avg_amt * 0.5 or 500))
    known_recipients = [r.lower() for r in profile.get("frequentRecipients", [])]

    is_known = bool(receiver) and any(
        receiver.lower() in k or k in receiver.lower() for k in known_recipients
    )

    # P4 -- Amount Outside Historical Range
    if amount > max_amt * 1.8 or (std_dev > 0 and amount > avg_amt + 3.5 * std_dev):
        factors.append({
            "name": "Severe Amount Deviation (P4)",
            "contribution": 28,
            "pattern": "P4",
            "description": (
                f"Rs.{amount:,.2f} is {amount/(avg_amt or 1):.1f}x your average "
                f"(Rs.{avg_amt:,.2f}) and exceeds your highest recorded amount."
            )
        })
        detected_patterns.append("P4")
    elif std_dev > 0 and amount > avg_amt + 2 * std_dev:
        factors.append({
            "name": "Elevated Transaction Amount (P4)",
            "contribution": 15,
            "pattern": "P4",
            "description": f"Rs.{amount:,.2f} noticeably exceeds typical spending variance (avg Rs.{avg_amt:,.2f})."
        })
        detected_patterns.append("P4")
    else:
        factors.append({
            "name": "Normal Amount Range",
            "contribution": 2,
            "pattern": "--",
            "description": f"Rs.{amount:,.2f} is consistent with typical range (avg Rs.{avg_amt:,.2f})."
        })

    # P1 -- Repeated Same-Amount Transactions (Rapid)
    if recent:
        same_amount_count = sum(
            1 for t in recent[:10] if abs(float(t.get("amount", -1)) - amount) < 1.0
        )
        if same_amount_count >= 2:
            contrib = min(25, 10 + same_amount_count * 5)
            factors.append({
                "name": f"Repeated Same-Amount Burst (P1) x{same_amount_count + 1}",
                "contribution": contrib,
                "pattern": "P1",
                "description": (
                    f"Rs.{amount:,.2f} was sent {same_amount_count + 1} times in quick succession. "
                    "Repeated identical amounts are a common indicator of automated fraud."
                )
            })
            detected_patterns.append("P1")

    # P2 -- Doubling/Multiplying Amount to Same Recipient
    if recent and receiver:
        same_recv_amounts = sorted(
            [float(t.get("amount", 0)) for t in recent[:8] if str(t.get("receiver", "")).lower() == receiver.lower()],
            reverse=True
        )
        if len(same_recv_amounts) >= 2:
            prev = same_recv_amounts[0]
            if prev > 0 and 1.7 <= (amount / prev) <= 2.5:
                factors.append({
                    "name": "Doubling Amount Pattern (P2)",
                    "contribution": 22,
                    "pattern": "P2",
                    "description": (
                        f"Amount to '{receiver}' roughly doubled: Rs.{prev:,.2f} to Rs.{amount:,.2f}. "
                        "Escalating amounts are a known fraud indicator."
                    )
                })
                detected_patterns.append("P2")

    # P3 -- Transaction Frequency Burst
    if recent:
        burst_count = len(recent[:15])
        daily_avg = float(profile.get("dailyAvgCount", 3))
        if burst_count >= max(5, daily_avg * 2.5):
            factors.append({
                "name": f"High Frequency Burst (P3) -- {burst_count} recent txns",
                "contribution": 20,
                "pattern": "P3",
                "description": (
                    f"Detected {burst_count} recent transactions vs. daily average of {daily_avg:.0f}. "
                    "A sudden burst may indicate account compromise."
                )
            })
            detected_patterns.append("P3")

    # P5 -- New Recipient + Unusual High Amount
    if not is_known and receiver and amount > avg_amt * 1.5:
        factors.append({
            "name": "New Recipient + High Amount (P5)",
            "contribution": 25,
            "pattern": "P5",
            "description": (
                f"'{receiver}' is a first-time recipient and Rs.{amount:,.2f} "
                f"is {amount/avg_amt:.1f}x your average -- high combined risk."
            )
        })
        detected_patterns.append("P5")
    elif not is_known and receiver:
        factors.append({
            "name": "New / First-Time Recipient",
            "contribution": 18,
            "pattern": "--",
            "description": f"'{receiver}' has never appeared in your verified transaction history."
        })
    else:
        factors.append({
            "name": "Known Frequent Recipient",
            "contribution": 0,
            "pattern": "--",
            "description": f"'{receiver}' matches a verified frequent contact."
        })

    # P6 -- New Recipient + Multiple Rapid Transactions
    if not is_known and recent:
        rapid_same_recv = sum(
            1 for t in recent[:6] if str(t.get("receiver", "")).lower() == receiver.lower()
        )
        if rapid_same_recv >= 2:
            factors.append({
                "name": f"New Recipient + Rapid Repeat (P6) x{rapid_same_recv + 1}",
                "contribution": 22,
                "pattern": "P6",
                "description": (
                    f"'{receiver}' is a new contact with {rapid_same_recv + 1} rapid consecutive "
                    "transactions -- strong indicator of social-engineering fraud."
                )
            })
            detected_patterns.append("P6")
    elif recent:
        same_recv_count = sum(
            1 for t in recent[:5] if str(t.get("receiver", "")).lower() == receiver.lower()
        )
        if same_recv_count >= 2 and "P6" not in detected_patterns:
            factors.append({
                "name": f"Rapid Repeated Transaction (P6) x{same_recv_count + 1}",
                "contribution": 18,
                "pattern": "P6",
                "description": f"Detected {same_recv_count + 1} transactions to '{receiver}' in a short window."
            })
            detected_patterns.append("P6")

    # P7 -- Multiple Different Recipients in Short Period
    if recent:
        unique_recent_receivers = set(
            str(t.get("receiver", "")).lower() for t in recent[:8] if t.get("receiver")
        )
        if len(unique_recent_receivers) >= 4:
            factors.append({
                "name": f"Multiple Recipients Burst (P7) -- {len(unique_recent_receivers)} different",
                "contribution": 18,
                "pattern": "P7",
                "description": (
                    f"Payments to {len(unique_recent_receivers)} different recipients in a short period. "
                    "Scatter pattern may indicate account takeover."
                )
            })
            detected_patterns.append("P7")

    # P8 -- Unusual Time of Day
    if hour < 5 or hour >= 23:
        factors.append({
            "name": "Late Night / Very Early Morning (P8)",
            "contribution": 18,
            "pattern": "P8",
            "description": f"Transaction at {time_str} is outside normal active hours (5 AM to 11 PM)."
        })
        detected_patterns.append("P8")
    elif hour < 7:
        factors.append({
            "name": "Early Morning Activity (P8)",
            "contribution": 8,
            "pattern": "P8",
            "description": f"Transaction at {time_str} is in an infrequent usage period (5-7 AM)."
        })

    # P9 -- Round/Structured Amount to Unknown Recipient
    round_values = {1000, 2000, 5000, 10000, 20000, 25000, 50000, 100000}
    if amount in round_values and not is_known:
        factors.append({
            "name": "Structured Round Amount to Unknown Recipient (P9)",
            "contribution": 12,
            "pattern": "P9",
            "description": (
                f"Exact round amount Rs.{amount:,.0f} sent to an unverified recipient. "
                "Common in structuring or test-fraud."
            )
        })
        detected_patterns.append("P9")

    # P10 -- Device Anomaly
    if device and "new device" in device.lower():
        factors.append({
            "name": "Unrecognized Device (P10)",
            "contribution": 12,
            "pattern": "P10",
            "description": f"Transaction from '{device}' -- device not previously seen in your profile."
        })
        detected_patterns.append("P10")

    # P11 -- Geographic/Location Anomaly
    if location and ("unknown" in location.lower() or "new location" in location.lower()):
        factors.append({
            "name": "Location Anomaly (P11)",
            "contribution": 10,
            "pattern": "P11",
            "description": f"Location '{location}' is unusual compared to your verified geography."
        })
        detected_patterns.append("P11")

    # P12 -- Sudden Large-Value Spike vs. Recent Average
    if recent and len(recent) >= 3:
        recent_amounts = [float(t.get("amount", 0)) for t in recent[:5] if float(t.get("amount", 0)) > 0]
        if recent_amounts:
            recent_avg = sum(recent_amounts) / len(recent_amounts)
            if recent_avg > 0 and amount > recent_avg * 3:
                factors.append({
                    "name": f"Sudden Value Spike (P12) -- {amount/recent_avg:.1f}x recent avg",
                    "contribution": 20,
                    "pattern": "P12",
                    "description": (
                        f"Rs.{amount:,.2f} is {amount/recent_avg:.1f}x your recent session average "
                        f"(Rs.{recent_avg:,.2f}). Sudden escalation is a key fraud signal."
                    )
                })
                detected_patterns.append("P12")

    # Aggregate
    raw_score = sum(f["contribution"] for f in factors)
    score = min(100, max(0, int(raw_score)))
    risk_info = get_risk_level(score)

    return _build_result(score, risk_info, detected_patterns, factors, profile.get("status", "Active"))


def _build_result(
    score: int,
    risk_info: Dict[str, Any],
    detected_patterns: List[str],
    factors: List[Dict[str, Any]],
    profile_status: str
) -> Dict[str, Any]:
    if score >= 80:
        risk_type = (
            "Combined High-Risk Pattern"
            if len(detected_patterns) > 1
            else (PATTERN_DESCRIPTIONS.get(detected_patterns[0], "Critical Anomaly") if detected_patterns else "Critical Anomaly")
        )
    elif score >= 60:
        risk_type = PATTERN_DESCRIPTIONS.get(detected_patterns[0], "High Anomaly Score") if detected_patterns else "High Anomaly Score"
    elif score >= 30:
        risk_type = PATTERN_DESCRIPTIONS.get(detected_patterns[0], "Moderate Variance") if detected_patterns else "Moderate Variance"
    else:
        risk_type = "Normal Transaction"

    enriched_patterns = [
        f"{pid}: {PATTERN_DESCRIPTIONS.get(pid, pid)}" for pid in detected_patterns
    ]

    return {
        "score": score,
        "level": risk_info["level"],
        "type": risk_type,
        "detected_patterns": enriched_patterns if enriched_patterns else ["No anomalies detected"],
        "pattern_ids": detected_patterns,
        "factors": factors,
        "profile_status": profile_status,
        "why": [f["description"] for f in factors if f.get("contribution", 0) > 0],
        "recommendation": generate_recommendation(score, factors)
    }


def generate_recommendation(score: int, factors: List[Dict[str, Any]]) -> str:
    if score < 30:
        return (
            "No significant suspicious indicators detected. "
            "Always confirm the recipient name before authorizing any payment."
        )
    elif score < 60:
        return (
            "Moderate behavioral variance detected. "
            "Verify the recipient identity and review the transaction details before completing the transfer."
        )
    elif score < 80:
        return (
            "High anomaly detected. Verify the recipient before proceeding. "
            "If you did not initiate this transaction, do NOT share your UPI PIN or OTP -- "
            "contact your bank through an official channel immediately."
        )
    else:
        return (
            "CRITICAL: Multiple severe fraud signals detected. Do NOT approve or continue. "
            "Independently verify this transaction and report to your bank's official fraud helpline immediately."
        )

File: frontend/src/engine/riskEngine.js
// ============================================
// UPI Sentinel AI — Risk Engine
// Deterministic, rule-based fraud risk scoring
// ============================================

// Configurable thresholds
export const RISK_THRESHOLDS = {
  LOW: { min: 0, max: 29, label: 'LOW', color: '#22c55e' },
  MEDIUM: { min: 30, max: 59, label: 'MEDIUM', color: '#f59e0b' },
  HIGH: { min: 60, max: 79, label: 'HIGH', color: '#ef4444' },
  CRITICAL: { min: 80, max: 100, label: 'CRITICAL', color: '#dc2626' },
};

export function getRiskLevel(score) {
  if (score <= RISK_THRESHOLDS.LOW.max) return RISK_THRESHOLDS.LOW;
  if (score <= RISK_THRESHOLDS.MEDIUM.max) return RISK_THRESHOLDS.MEDIUM;
  if (score <= RISK_THRESHOLDS.HIGH.max) return RISK_THRESHOLDS.HIGH;
  return RISK_THRESHOLDS.CRITICAL;
}

export function getRiskColor(score) {
  return getRiskLevel(score).color;
}

export function analyseTransaction(transaction, profile, recentTransactions = []) {
  const factors = [];
  let totalScore = 0;

  // 1. Unusual Amount
  const amountFactor = checkUnusualAmount(transaction, profile);
  factors.push(amountFactor);
  totalScore += amountFactor.contribution;

  // 2. New Recipient
  const recipientFactor = checkNewRecipient(transaction, profile);
  factors.push(recipientFactor);
  totalScore += recipientFactor.contribution;

  // 3. Unusual Time
  const timeFactor = checkUnusualTime(transaction, profile);
  factors.push(timeFactor);
  totalScore += timeFactor.contribution;

  // 4. Rapid Repeated Transaction
  const rapidFactor = checkRapidRepeated(transaction, recentTransactions);
  factors.push(rapidFactor);
  totalScore += rapidFactor.contribution;

  // 5. Frequency Anomaly
  const frequencyFactor = checkFrequencyAnomaly(transaction, recentTransactions, profile);
  factors.push(frequencyFactor);
  totalScore += frequencyFactor.contribution;

  // 6. Repeated Amount Pattern
  const amountPatternFactor = checkRepeatedAmountPattern(transaction, recentTransactions);
  factors.push(amountPatternFactor);
  totalScore += amountPatternFactor.contribution;

  // 7. Behaviour Deviation
  const deviationFactor = checkBehaviourDeviation(transaction, profile);
  factors.push(deviationFactor);
  totalScore += deviationFactor.contribution;

  // 8. Device Anomaly
  const deviceFactor = checkDeviceAnomaly(transaction, profile);
  factors.push(deviceFactor);
  totalScore += deviceFactor.contribution;

  // 9. Location Anomaly
  const locationFactor = checkLocationAnomaly(transaction, profile, recentTransactions);
  factors.push(locationFactor);
  totalScore += locationFactor.contribution;

  // 10. Combined Risk Multiplier
  const activeFacts = factors.filter(f => f.contribution > 0);
  if (activeFacts.length >= 3) {
    const combinedBonus = Math.min(Math.floor(activeFacts.length * 2.5), 15);
    factors.push({
      name: 'Combined Risk Multiplier',
      contribution: combinedBonus,
      description: `${activeFacts.length} risk indicators detected together increase overall risk`,
      severity: 'high',
    });
    totalScore += combinedBonus;
  }

  // Cap at 100
  totalScore = Math.min(totalScore, 100);

  const riskLevel = getRiskLevel(totalScore);
  const riskType = determineRiskType(factors, totalScore);
  const recommendation = generateRecommendation(riskLevel.label, factors);

  return {
    score: totalScore,
    level: riskLevel.label,
    levelInfo: riskLevel,
    type: riskType,
    factors: factors.filter(f => f.contribution > 0).sort((a, b) => b.contribution - a.contribution),
    allFactors: factors,
    recommendation,
  };
}

function checkUnusualAmount(txn, profile) {
  const amount = txn.amount;
  const avg = profile.averageAmount || 1420;
  const max = profile.maxAmount || 12000;
  const stdDev = profile.amountStdDev || 800;

  let contribution = 0;
  let description = '';

  const deviation = Math.abs(amount - avg) / (stdDev || 1);

  if (amount > max * 2) {
    contribution = 24;
    description = `₹${amount.toLocaleString()} far exceeds normal range (avg ₹${avg.toLocaleString()})`;
  } else if (amount > max * 1.5) {
    contribution = 18;
    description = `₹${amount.toLocaleString()} significantly exceeds normal behaviour`;
  } else if (deviation > 3) {
    contribution = 15;
    description = `₹${amount.toLocaleString()} is ${deviation.toFixed(1)} standard deviations from average`;
  } else if (deviation > 2) {
    contribution = 10;
    description = `₹${amount.toLocaleString()} is somewhat above normal range`;
  } else if (deviation > 1.5) {
    contribution = 5;
    description = `₹${amount.toLocaleString()} is slightly above typical range`;
  } else {
    description = `₹${amount.toLocaleString()} is within normal range (avg ₹${avg.toLocaleString()})`;
  }

  return { name: 'Unusual Amount', contribution, description, severity: contribution > 15 ? 'high' : contribution > 5 ? 'medium' : 'low' };
}

function checkNewRecipient(txn, profile) {
  const recipients = profile.frequentRecipients || [];
  const isKnown = recipients.some(
    r => r.name.toLowerCase() === (txn.receiver || '').toLowerCase() ||
         r.upiId === txn.receiverUpi
  );

  if (isKnown) {
    return { name: 'Known Recipient', contribution: 0, description: `${txn.receiver} is a known recipient`, severity: 'low' };
  }

  return {
    name: 'New Recipient',
    contribution: 18,
    description: `${txn.receiver} has never appeared in transaction history`,
    severity: 'medium',
  };
}

function checkUnusualTime(txn, profile) {
  const hour = txn.hour;
  const normalStart = profile.normalTimeStart || 8;
  const normalEnd = profile.normalTimeEnd || 22;

  if (hour >= normalStart && hour <= normalEnd) {
    return { name: 'Normal Time', contribution: 0, description: `${txn.time} is within normal hours (${normalStart}:00 – ${normalEnd}:00)`, severity: 'low' };
  }

  // Late night / early morning — higher risk
  const hoursOutside = hour < normalStart ? normalStart - hour : hour - normalEnd;
  const contribution = Math.min(hoursOutside * 4, 18);

  return {
    name: 'Unusual Time',
    contribution,
    description: `${txn.time} is outside normal transaction hours (${normalStart}:00 AM – ${normalEnd}:00 PM)`,
    severity: contribution > 10 ? 'high' : 'medium',
  };
}

function checkRapidRepeated(txn, recentTransactions) {
  if (!recentTransactions || recentTransactions.length === 0) {
    return { name: 'Rapid Repeated Transaction', contribution: 0, description: 'No rapid repeated pattern detected', severity: 'low' };
  }

  // Find transactions to same recipient within 5 minutes
  const sameRecipient = recentTransactions.filter(t => {
    const sameReceiver = (t.receiver || '').toLowerCase() === (txn.receiver || '').toLowerCase();
    const sameDate = t.dateISO === txn.dateISO;
    if (!sameReceiver || !sameDate) return false;

    const timeDiff = Math.abs((txn.hour * 60 + txn.minute) - (t.hour * 60 + t.minute));
    return timeDiff <= 5;
  });

  if (sameRecipient.length >= 2) {
    return {
      name: 'Rapid Repeated Transaction',
      contribution: 22,
      description: `${sameRecipient.length + 1} transactions to same recipient within minutes`,
      severity: 'high',
    };
  } else if (sameRecipient.length === 1) {
    return {
      name: 'Rapid Repeated Transaction',
      contribution: 12,
      description: `2 transactions to same recipient within minutes`,
      severity: 'medium',
    };
  }

  return { name: 'Rapid Repeated Transaction', contribution: 0, description: 'No rapid repeated pattern detected', severity: 'low' };
}

function checkFrequencyAnomaly(txn, recentTransactions, profile) {
  const avgDaily = profile.averageDailyTransactions || 4;
  const todayTxns = recentTransactions.filter(t => t.dateISO === txn.dateISO);
  const todayCount = todayTxns.length + 1;

  if (todayCount > avgDaily * 3) {
    return {
      name: 'Frequency Anomaly',
      contribution: 15,
      description: `${todayCount} transactions today, normal is ~${avgDaily}/day`,
      severity: 'high',
    };
  } else if (todayCount > avgDaily * 2) {
    return {
      name: 'Frequency Anomaly',
      contribution: 8,
      description: `${todayCount} transactions today, somewhat above normal (~${avgDaily}/day)`,
      severity: 'medium',
    };
  }

  return { name: 'Normal Frequency', contribution: 0, description: `Transaction frequency is consistent with profile`, severity: 'low' };
}

function checkRepeatedAmountPattern(txn, recentTransactions) {
  if (!recentTransactions || recentTransactions.length < 2) {
    return { name: 'Amount Pattern', contribution: 0, description: 'No repeated amount pattern', severity: 'low' };
  }

  const sameAmount = recentTransactions.filter(t =>
    t.amount === txn.amount && t.dateISO === txn.dateISO
  );

  if (sameAmount.length >= 2) {
    return {
      name: 'Repeated Amount Pattern',
      contribution: 12,
      description: `₹${txn.amount.toLocaleString()} repeated ${sameAmount.length + 1} times today`,
      severity: 'medium',
    };
  } else if (sameAmount.length === 1) {
    return {
      name: 'Repeated Amount Pattern',
      contribution: 5,
      description: `₹${txn.amount.toLocaleString()} appears twice today`,
      severity: 'low',
    };
  }

  return { name: 'Amount Pattern', contribution: 0, description: 'No repeated amount pattern', severity: 'low' };
}

function checkBehaviourDeviation(txn, profile) {
  let deviationCount = 0;

  // Amount check
  if (txn.amount > (profile.maxAmount || 12000) * 1.5) deviationCount++;
  // Time check
  if (txn.hour < (profile.normalTimeStart || 8) || txn.hour > (profile.normalTimeEnd || 22)) deviationCount++;
  // Recipient check
  const knownRecipients = profile.frequentRecipients || [];
  const isKnown = knownRecipients.some(r => r.name.toLowerCase() === (txn.receiver || '').toLowerCase());
  if (!isKnown) deviationCount++;

  if (deviationCount >= 3) {
    return {
      name: 'Behaviour Deviation',
      contribution: 15,
      description: 'Overall behaviour significantly deviates from established profile',
      severity: 'high',
    };
  } else if (deviationCount === 2) {
    return {
      name: 'Behaviour Deviation',
      contribution: 10,
      description: 'Transaction shows moderate deviation from behaviour profile',
      severity: 'medium',
    };
  } else if (deviationCount === 1) {
    return {
      name: 'Behaviour Deviation',
      contribution: 5,
      description: 'Minor deviation from behaviour profile',
      severity: 'low',
    };
  }

  return { name: 'Behaviour Match', contribution: 0, description: 'Overall behaviour matches profile', severity: 'low' };
}

function checkDeviceAnomaly(txn, profile) {
  if (!txn.device) {
    return { name: 'Device Info', contribution: 0, description: 'Device data unavailable', severity: 'low' };
  }

  const knownDevices = profile.knownDevices || ['Samsung Galaxy S23', 'iPhone 14'];
  const isKnown = knownDevices.some(d => d.toLowerCase() === txn.device.toLowerCase());

  if (!isKnown) {
    return {
      name: 'New Device',
      contribution: 10,
      description: `${txn.device} has not been seen before`,
      severity: 'medium',
    };
  }

  return { name: 'Known Device', contribution: 0, description: `${txn.device} is a known device`, severity: 'low' };
}

function checkLocationAnomaly(txn, profile, recentTransactions) {
  if (!txn.location) {
    return { name: 'Location Info', contribution: 0, description: 'Location data unavailable', severity: 'low' };
  }

  const usualLocation = profile.usualLocation || 'Chennai';
  if (txn.location.toLowerCase() === usualLocation.toLowerCase()) {
    return { name: 'Normal Location', contribution: 0, description: `Transaction from ${txn.location} (usual location)`, severity: 'low' };
  }

  // Check for impossible travel
  if (recentTransactions.length > 0) {
    const lastTxn = recentTransactions[0];
    if (lastTxn.location && lastTxn.location.toLowerCase() !== txn.location.toLowerCase()) {
      const timeDiff = Math.abs((txn.hour * 60 + txn.minute) - (lastTxn.hour * 60 + lastTxn.minute));
      if (timeDiff < 30 && lastTxn.dateISO === txn.dateISO) {
        return {
          name: 'Location Anomaly',
          contribution: 15,
          description: `Potential impossible-travel: ${lastTxn.location} to ${txn.location} in ${timeDiff} minutes`,
          severity: 'high',
        };
      }
    }
  }

  return {
    name: 'Location Anomaly',
    contribution: 8,
    description: `Transaction from ${txn.location}, usual location is ${usualLocation}`,
    severity: 'medium',
  };
}

function determineRiskType(factors, score) {
  const activeFactors = factors.filter(f => f.contribution > 0).map(f => f.name);

  if (activeFactors.includes('Rapid Repeated Transaction') && score >= 60) {
    return 'Repeated Rapid Transaction';
  }
  if (activeFactors.includes('Location Anomaly') && activeFactors.includes('Unusual Time')) {
    return 'Velocity / Frequency Anomaly';
  }
  if (activeFactors.length >= 3 && score >= 80) {
    return 'Combined High-Risk Pattern';
  }
  if (activeFactors.includes('Unusual Amount') && factors.find(f => f.name === 'Unusual Amount')?.contribution >= 18) {
    return 'Unusual Amount Pattern';
  }
  if (activeFactors.includes('New Recipient') && activeFactors.includes('Unusual Time')) {
    return 'Potential Scam Pattern';
  }
  if (activeFactors.includes('Frequency Anomaly')) {
    return 'Velocity / Frequency Anomaly';
  }
  if (activeFactors.includes('Unusual Time')) {
    return 'Time-Based Behaviour Anomaly';
  }
  if (activeFactors.includes('New Recipient')) {
    return 'New Recipient Risk';
  }
  if (activeFactors.includes('Behaviour Deviation')) {
    return 'Behavioural Deviation';
  }
  if (score <= 29) {
    return 'Normal Transaction';
  }
  return 'General Risk Pattern';
}

function generateRecommendation(level, factors) {
  const activeFactorNames = factors.filter(f => f.contribution > 0).map(f => f.name);

  if (level === 'LOW') {
    return 'No significant suspicious indicators were detected by the current prototype.';
  }

  if (level === 'MEDIUM') {
    return 'Some unusual indicators were detected. Review the transaction details before proceeding.';
  }

  const recommendations = ['Verify the recipient.'];

  if (activeFactorNames.includes('New Recipient') || activeFactorNames.includes('Unusual Amount')) {
    recommendations.push('Do not approve an unfamiliar transaction.');
  }

  recommendations.push('If you did not initiate the transaction, contact your bank/payment provider through an official channel.');
  recommendations.push('Review recent account activity.');

  if (activeFactorNames.includes('Rapid Repeated Transaction')) {
    recommendations.unshift('Rapid repeated transaction pattern detected.');
  }

  return recommendations.join(' ');
}

export default {
  analyseTransaction,
  getRiskLevel,
  getRiskColor,
  RISK_THRESHOLDS,
};
